projects/shared/src/lib/services/bookmark.service.ts

```diff
@@ -0,0 +1,270 @@
+   *   relied on, which is discouraged by this.
+   * - The xorshift codec is derived from two things: the fragment's ID,
+   *   and its length. This ensures the fragment's values can't be accidentally
+   *   or purposefully transposed with another fragment's.
+   * - As another modification, the byte being encoded is xor'd into the xorshift
+   *   state, causing a small cascade effect. This, together with the extra 0 byte
+   *   being encoded, functions as a very slight checksum, being able to handle at
+   *   least one typo (if anyone considers these typable to begin with).
+   *
+   * All of these are necessary to increase the robustness of the format:
+   * - The ID number and length protect typos in the fragment id
+   * - The cipher feedback ensures typos inside the content are detected
+   * - The extra 0 byte and cipher feedback ensure the data is probably untouched
+   *
+   * As an added benefit, the obfuscation lightly decreases the chance of getting sensible
+   * words out of the whole thing.
+   *
+   * Of course, this isn't a perfect protection, but almost all one-character typos will
+   * be caught by this, which seems good enough.
+   */
+  private static encodeBytes(id: string, data: Uint8Array): string {
+    if (data.length === 0) {
+      return '';
+    }
+
+    // encode the length of the data to encode, as well as the fragment's ID, into the
+    // PRNG. This uses two primes, slightly distant from each other, chosen by fair dice roll.
+    const obfuscation = new Xorshift(data.length * 17 + ([...id].reduce((b, a) => b + a.charCodeAt(0), 0)) * 257);
+    const cipheredBytes = [...data].map(a => obfuscation.encode(a));
+
+    // Encode a single 0 byte at the end. This serves as the "checksum", lowering the chances
+    // of typos being picked up on.
+    cipheredBytes.push(obfuscation.encode(0) & 0xFF);
+
+    // Encode cipheredBytes as base64url.
+    return btoa(String.fromCharCode.apply(null, cipheredBytes)).replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_');
+  }
+
+  private static decodeBytes(id: string, encoded: string): Uint8Array | undefined {
+    const cipheredBytes = atob(encoded.replace(/-/g, '+').replace(/_/g, '/'));
+    // minus one, as cipheredBytes has one extra "0" byte, which isn't counted in the
+    // length used for obfuscation.
+    const obfuscation = new Xorshift((cipheredBytes.length - 1) * 17 + ([...id].reduce((b, a) => b + a.charCodeAt(0), 0)) * 257);
+
+    const data = Uint8Array.from(cipheredBytes, c => obfuscation.decode(c.charCodeAt(0)));
+
+    // verify the checksum
+    if (data[data.length - 1] !== 0) {
+      return undefined;
+    }
+
+    return new Uint8Array(data.buffer, 0, data.length - 1);
+  }
+
+  private pendingFragments: Map<string, string> = new Map();
+  private fragments: Map<BookmarkFragmentProtoDescriptor | BookmarkFragmentStringDescriptor, BookmarkFragmentInfo> = new Map();
+  private joinedBookmark: Subject<string | undefined> = new Subject();
```

Comment:
If this is a `BehaviorSubject` the URI fragment would immediately be emptied upon loading, which means the application would need logic to prevent that; hence being a plain `Subject`.